product-scenarios/3-securing-an-api-with-different-security-protocols/3.2-api-developers-want-to-secure-their-api-using-an-user-access-control-machanism/3.2.1-api-developer-wants-to-secure-api-resources-using-user-roles/src/test/java/org/wso2/am/scenario/tests/api/secure/userRoles/SecureUsingUserRolesNegativeTestCase.java

```diff
@@ -116,23 +166,77 @@ private void deleteRoles() throws APIManagementException {
+        accessToken = generateKeysAndAccessToken(CUSTOMER, CUSTOMER_LOGIN_PW, ITEM_VIEW);
+
+        //Update the Role in scope
+        updateAPISwagger(AGENT_ROLE, swaggerJson);
+
+        //Test invocation of related resource with old token
+        requestHeaders.put("Authorization", "Bearer " + accessToken);
+        response = HttpRequestUtil
+                .doGet(gatewayHttpsURL + "/contextJsonV2/" + apiVersion, requestHeaders);
+        assertNotEquals(response.getResponseCode(), Response.Status.OK.getStatusCode(),
+                "user can access the GET Method using previous scope token");
+
+        //Reset the API to initial state
+        updateAPISwagger(MANAGER_ROLE, swaggerJson);
+        apiStore.removeAPISubscriptionByName(apiName, apiVersion, SUPER_USER, applicationName);
+        apiStore.removeApplication(applicationName);
+    }
+
+    @Test(description = "3.2.1.7")
+    public void testResourceWithConfiguredRoleBasedScopeInvokeByInvalidUser() throws Exception {
+        //Adding a scope with two roles
+        Map<String, String> requestHeaders;
+        HttpResponse response;
+        requestHeaders = new HashMap<String, String>();
+        String accessToken;
+
+        apiStore.login(AGENT, AGENT_LOGIN_PW);
+        createNewApplicationAndSubscribe(AGENT, AGENT_LOGIN_PW);
+        accessToken = generateKeysAndAccessToken(AGENT, AGENT_LOGIN_PW, ITEM_VIEW);
+
+        requestHeaders.put("Authorization", "Bearer " + accessToken);
+        response = HttpRequestUtil
+                .doGet(gatewayHttpsURL + "/contextJsonV2/" + apiVersion, requestHeaders);
+        assertNotEquals(response.getResponseCode(), Response.Status.OK.getStatusCode(),
+                "Invalid user can access the GET Method");
+        apiStore.removeAPISubscriptionByName(apiName, apiVersion, AGENT, applicationName);
+        apiStore.removeApplication(applicationName);
+    }
+
+    @Test(description = "3.2.1.11")
+    public void testInvokeResourceByTokenWithoutRequiredRole() throws Exception {
+        //Adding a scope with two roles
+        Map<String, String> requestHeaders;
+        HttpResponse response;
+        requestHeaders = new HashMap<String, String>();
+        String accessToken;
+
+        apiStore.login(SUPER_USER, SUPER_USER_LOGIN_PW);
+        createNewApplicationAndSubscribe(SUPER_USER, SUPER_USER_LOGIN_PW);
+        accessToken = generateKeysAndAccessToken(SUPER_USER, SUPER_USER_LOGIN_PW, "");
+
+        requestHeaders.put("Authorization", "Bearer " + accessToken);
+        response = HttpRequestUtil
+                .doGet(gatewayHttpsURL + "/contextJsonV2/" + apiVersion, requestHeaders);
+        assertNotEquals(response.getResponseCode(), Response.Status.OK.getStatusCode(),
+                "user can access the GET Method by token which ");
+        apiStore.removeAPISubscriptionByName(apiName, apiVersion, SUPER_USER, applicationName);
+        apiStore.removeApplication(applicationName);
```

Comment:
Better to create a new separate application for each test case so that you dont need to delete them and add them back in the next test case. Delete all applications created in the destroy() method at the end, as it has been done for users. This will remove unnecessary boiler plate code from individual test cases